core/src/sumeragi/mod.rs

```diff
@@ -250,9 +251,38 @@ impl Sumeragi {
     pub fn initialize_and_start_thread(
         sumeragi: Arc<Self>,
         genesis_network: Option<GenesisNetwork>,
+        block_hashes: Vec<HashOf<VersionedCommittedBlock>>,
     ) -> ThreadHandler {
         let wsv = sumeragi.wsv_mutex_access().clone();
 
+        for (i, block_hash) in block_hashes
+            .iter()
+            .take(block_hashes.len().saturating_sub(1))
+            .enumerate()
+        {
+            let block_height: u64 = (i + 1).try_into().unwrap();
+            let block_ref = sumeragi.internal.kura.get_block_by_height(block_height).expect("Sumeragi could not load block that was reported as present. Please check that the block storage was not disconnected.");
+            assert_eq!(
+                block_ref.hash(),
+                *block_hash,
+                "Kura init correctly reported the block hash."
+            );
+
+            wsv.apply(&block_ref)
+                .expect("Failed to apply block to wsv in init.");
+        }
+        let delayed_wsv = wsv.clone();
+
+        if let Some(block_hash) = block_hashes.last() {
+            let block_ref = sumeragi.internal.kura.get_block_by_hash(block_hash).expect("Sumeragi could not load block that was reported as present. Please check that the block storage was not disconnected.");
```

Comment:
You can get block by height here as well, no?

Reply:
I was hesitant to do that originally but you are right it is better that way.